kratos/containers/model.cpp

```diff
@@ -303,18 +303,25 @@ void Model::save(Serializer& rSerializer) const
 
 void Model::load(Serializer& rSerializer)
 {
-    //we construct auxiliary arrays to avoid having to serialize sets and maps of unique_ptrs
+    // We construct auxiliary arrays to avoid having to serialize sets and maps of unique_ptrs
     std::vector<std::string> aux_names;
 
+    // Serialize the model part names and the data value container
     rSerializer.load("ModelPartNames", aux_names);
     rSerializer.load("ModelData", mDataValueContainer);
 
+    // Generate the model parts
     for(IndexType i=0; i<aux_names.size(); ++i) {
         //NOTE: CreateModelPart CANNOT be used here
         auto dummy_list = Kratos::make_intrusive<VariablesList>();
-        ModelPart* pmodel_part = new ModelPart(aux_names[i], 1, dummy_list, *this );
-        rSerializer.load(aux_names[i], pmodel_part);
-        mRootModelPartMap.insert(std::make_pair(aux_names[i],std::unique_ptr<ModelPart>(pmodel_part)));
+        const std::string& r_name = aux_names[i];
```

Comment:
can you explain why the old version was wrong? i simply cannot see it...

Reply:
Me neither, but this solves the issue...

Looks like lifespan issue, but should not be a problem IMHO.

Reply:
I can try to look for a failing problem, but I cannot find one without sharing an industrial case. The only thing is that it fails when more than one model part, and the model part empty is the last one.

Reply:
(Only in Windows BTW)

Reply:
I was looking to this last night and I don't know what to say. There are to many things happening at the same time that are different from the original to the new code, so I will go change by change:

1 - Using a `const std::string &` to the best of my knowledge is syntax sugar, the code produced by the compiler if that were the only change (to understand):

```c++
const std::string& r_name = aux_names[i];
ModelPart* pmodel_part = new ModelPart(r_name, 1, dummy_list, *this );

rSerializer.load(r_name, pmodel_part);
mRootModelPartMap.insert(std::make_pair(r_name,std::unique_ptr<ModelPart>(pmodel_part)));
```

should be exactly the same.

2 - Inverting the insert / load order may cause some behavior to change (specially if there are pointers involved in the load call), so I can imagine that may help to solve the problem:

```c++
mRootModelPartMap.insert(std::make_pair(r_name,std::unique_ptr<ModelPart>(pmodel_part)));
// Pseudo 
rSerializer.load(r_name, mRootModelPartMap.get(r_name));
```

3 - Manually loading all the `ModelParts` in the auxiliary list not following the order in which they were serialized (which we should safely assume is the same in which they appear in the `aux_names` but at this point who knows...) does not guarantee we are doing the same.

```c++
for (name in aux_names) {
    // Create
    // Insert
}

for (name in aux_names) {
    // Load
}
```

Also, with this change made, I can se a corner case where a ModelPart that has a repeated name for some reason would fail to be loaded in the original code, but work in the second one, just because the `load` function from the serializer will load it on the top of an already existing Modelpart that were created by the original, so that would also "solve" a possible crash.

TL;DR:  I doubt that the alias for the name does nothing unless some optimization dark magic from the compiler is changing things, and from the other changes is impossible to tell without having more details. If it works I see no harm in merging but I would like to know what exactly is happening.

Reply:
> 2 - Inverting the insert / load order may cause some behavior to change (specially if there are pointers involved in the load call), so I can imagine that may help to solve the problem:

This is the one I'd bet on. The only difference I see between the old and new versions is that the `ModelPart` tree is complete before any `ModelPart`s start getting loaded, so any object that would want to store some reference/pointer to a specific `ModelPart` can safely do so (not a great idea tho).

The thing that worries me about our deserialization system is potentially loading data that is used for keys in a hash/ordered table **after** the object is constructed. Example:
1) construct a `ModelPart` and insert it into the hash table of `Model`
2) deserialize `ModelPart`
    - while deserializing, the `ModelPart`'s name also gets deserialized, which is used for computing its hash in the hash table

Thankfully, step 2 is not a problem right now because we require that the deserialized name is identical to the `ModelPart`'s name we inserted it with. However, if we a similar situation somewhere else, and the name/id/etc **does** get overwritten, that's a disaster.

##

@loumalouomega can you compile with address sanitizer and run your case with that? I usually compile in `RelWithDebInfo` mode and pass `-fsanitize=address` to `CMAKE_CXX_FLAGS`. Before running your case, you'll need to load the address sanitizer lib (`export LD_PRELOAD=/usr/lib(libasan.so` or wherever your `libasan.so` is).

Reply:
Sorry, after epiphany I don't have acess to a computer